dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+        -------
+        val : float
+            the negative mutual information of the input images after warping
+            the moving image by the currently set transform with `xopt`
+            parameters
+        """
+        if self.samples is None:
+            self._update_dense(xopt, False)
+        else:
+            self._update_sparse(xopt, False)
+        return -1 * self.metric_val
+
+    def gradient(self, xopt):
+        r""" Numeric value of the metric's gradient at the given parameters
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        grad : array, shape(n,)
+            the gradient of the negative Mutual Information
+        """
+        if self.samples is None:
+            self._update_dense(xopt, True)
+        else:
+            self._update_dense(xopt, True)
+        return self.metric_grad * (-1)
+
+    def value_and_gradient(self, xopt):
+        r""" Numeric value of the metric and its gradient at given parameters
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        val : float
+            the negative mutual information of the input images after warping
+            the moving image by the currently set transform with `xopt`
+            parameters
+        grad : array, shape(n,)
+            the gradient of the negative Mutual Information
+        """
+        if self.samples is None:
+            self._update_dense(xopt, True)
+        else:
+            self._update_sparse(xopt, True)
+        return -1 * self.metric_val, self.metric_grad * (-1)
+
+
+class AffineRegistration(object):
+    def __init__(self,
+                 metric=None,
+                 level_iters=None,
+                 sigmas=None,
+                 factors=None,
+                 method='L-BFGS-B',
+                 ss_sigma_factor=None,
+                 options=None):
+        r""" Initializes an instance of the AffineRegistration class
+
+        Parameters
+        ----------
+        metric : object, optional
+            an instance of a metric. The default is None, implying
+            the Mutual Information metric with default settings.
+        level_iters : list, optional
+            the number of iterations at each level of the Gaussian pyramid.
+            `level_iters[0]` corresponds to the coarsest level,
+            `level_iters[-1]` the finest, where n is the length of the list.
+            By default, a 3-level Gaussian pyramid with iterations list
```

Comment:
> Is Gaussian pyramid the same as scale space? 
I don't know much about scale space theory either. What we have is just a sequence of images smoothed with Gaussian kernels with different sigmas (a "scale-space" representation of each input image, using a Gaussian kernel in particular). In practice, these images are, at some point, sub-sampled so that they can be processed more efficiently. For me, this set of smoothed-and-subsampled images form a "Gaussian pyramid" (you can imagine the smallest image at the top of the pyramid). The ScaleSpace class does not subsample the image, it just smooths it out, that's why I didn't like that name "Gaussian pyramid". 

Anyway, in summary, for me the only difference is the subsampling part and the fact that the Gaussian pyramid uses a Gaussian kernel for smoothing, while the scale space does not involve any subsampling, and it may use any smoothing kernel. That's why, for me "level" and "scale" refer to the same thing. I just changed "Gaussian pyramid" to "scale space" everywhere.